File: 0-event-ingestion-kafka/ingest-api/0-typescript/src/events/publish-click.dto.ts
import { IsNotEmpty, IsObject, IsOptional, IsString } from "class-validator"

/**
 * Request body for `POST /clicks`.
 *
 * `clickId` is the idempotency key: the consumer uses it to drop duplicates,
 * so the producer treats it as a first-class field rather than free-form data.
 */
export class PublishClickDto {
    /**
     * Client-generated unique id for this click (idempotency key).
     *
     * Re-sending the same `clickId` (retry, double-fire) must be counted once
     * by the consumer. Optional here so the API can mint one when omitted.
     */
    @IsString()
    @IsOptional()
    clickId?: string

    /** Ad identifier the click belongs to; also used as the Kafka partition key. */
    @IsString()
    @IsNotEmpty()
    adId!: string

    /** Arbitrary contextual payload (device, geo, referrer...). */
    @IsObject()
    @IsOptional()
    payload?: Record<string, unknown>
}

File: 0-event-ingestion-kafka/ingest-api/1-java/src/main/java/com/starci/ingest/PublishClickDto.java
package com.starci.ingest;

import java.util.Map;

/**
 * Request body for {@code POST /clicks}.
 *
 * <p>{@code clickId} is the idempotency key: re-sending the same {@code clickId}
 * (retry, double-fire) must be counted once by the consumer, so the producer
 * treats it as a first-class field rather than free-form data. It is optional so
 * the API can mint one when omitted. {@code adId} is required and also serves as
 * the Kafka partition key. {@code payload} is arbitrary contextual data.
 */
public class PublishClickDto {
    public String clickId;
    public String adId;
    public Map<String, Object> payload;
}
